### PR TITLE
[Fix #95] robust handling of DOIs

### LINKED ISSUE #95: robust handling of DOIs
Currently, only `pubs add -D 10.1007/s00422-012-0514-6` works. But given how DOIs are frequently displayed in webpages and articles, pubs should also accept:

```
pubs add -D https://doi.org/10.1007/s00422-012-0514-6
pubs add -D http://doi.org/10.1007/s00422-012-0514-6
pubs add -D doi.org/10.1007/s00422-012-0514-6
```

### PR DESCRIPTION
**Pull request by wflynny**
Added DOI regex function to utils.py which is called in add_cmd.py
upon specifying a new DOI.  DOI validation applies directly on
argument parsing by using a custom argparse.Action.

Two concerns:  

1.  One thing that I don't like is that I imported `argparse` into `add_cmd.py` in order to set up the custom `argparse.Action` to validate DOIs at parse time.  
2.  Does the DOI parsing need to happen anywhere else?  I thought about adding it to `apis.doi2bibtex` but that is only ever called in `add_cmd.py`.

### CODE REVIEW

File: pubs/commands/add_cmd.py [hunk 1]
```diff
@@ -7,13 +8,21 @@
 from .. import apis
 from .. import color
 from .. import pretty
+from .. import utils
```
Comment (omangin) [inline pubs/commands/add_cmd.py]: PEP8: _Surround top-level function and class definitions with two blank lines._

File: pubs/commands/add_cmd.py [hunk 2]
```diff
@@ -7,13 +8,21 @@
...
 
+class ValidateDOI(argparse.Action):
+    def __call__(self, parser, namespace, values, option_string=None):
+        doi = values
+        new_doi = utils.standardize_doi(doi)
+        if not new_doi:
+            raise ValueError("Not a valid doi: %s", doi)
```
Comment (omangin) [inline pubs/commands/add_cmd.py]: I would move this (raising the exception) to `standardize_doi`.

File: pubs/utils.py [hunk 3]
```diff
@@ -44,3 +46,36 @@ def resolve_citekey_list(repo, citekeys, ui=None, exit_on_fail=True):
...
+        re.compile(r'(10\.1207/[\w\d]+\&\d+_\d+)')
+        )
+
+    for doi_regex in doi_regexes:
+        match = doi_regex.search(doi)
+        if match:
+            new_doi = match.group(0)
+            break
+    else:
+        new_doi = None
```
Comment (omangin) [inline pubs/utils.py]: This should raise the `ValueError` exception you have in the other file.

File: pubs/utils.py [hunk 4]
```diff
@@ -44,3 +46,36 @@ def resolve_citekey_list(repo, citekeys, ui=None, exit_on_fail=True):
...
+    https://www.doi.org/doi_handbook/2_Numbering.html
+    https://www.crossref.org/blog/dois-and-matching-regular-expressions/"""
+    """ :returns standardized doi """
+    doi_regexes = (
+        re.compile(r'(10\.\d{4,9}/[-._;()/:A-z0-9\>\<]+)'),
+        re.compile(r'(10.1002/[^\s]+)'),
+        re.compile(r'(10\.\d{4}/\d+-\d+X?(\d+)\d+<[\d\w]+:[\d\w]*>\d+.\d+.\w+;\d)'),
+        re.compile(r'(10\.1021/\w\w\d+\+)'),
+        re.compile(r'(10\.1207/[\w\d]+\&\d+_\d+)')
+        )
```
Comment (omangin) [inline pubs/utils.py]: Can't you join all these regexp with `|` operators? That way we can also get rid of the for loop below.

File: pubs/utils.py [hunk 5]
```diff
@@ -44,3 +46,36 @@ def resolve_citekey_list(repo, citekeys, ui=None, exit_on_fail=True):
...
+    -   http://doi.org/<doi>
+    -   https://dx.doi.org/<doi>
+    -   http://dx.doi.org/<doi>
+    -   dx.doi.org/<doi>
+    -   doi.org/<doi>
+    and attempts to verify doi adherence to DOI handbook standards and
+    crossref.org advice:
+    https://www.doi.org/doi_handbook/2_Numbering.html
+    https://www.crossref.org/blog/dois-and-matching-regular-expressions/"""
+    """ :returns standardized doi """
```
Comment (omangin) [inline pubs/utils.py]: Shouldn't this be in the same multi-line string?

**Comment by wflynny** [comment]
Thanks @omangin, great comments.  I think I got everything.  Let me know if I missed anything.

**Comment by omangin** [comment]
Perfect. Thanks again @wflynny!

**Pull request merged**